<!--
 %\VignetteEngine{knitr::knitr}
-->

---
Author: Jean-Philippe Fortin, Kasper Daniel Hansen
Date: July 13, 2015
---

<!-- rmarkdown v1 -->
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{450k methylation workflow with minfi}
-->

<!-- to compile this: library("knitr"); knit2html("methylation450k.Rmd") -->

<!--
     # a list of all required libraries:
     reqlibs = sub(".*library\\(\"(.*?)\"\\).*","\\1",grep("library\\(",readLines("methylation450k.Rmd"),value=TRUE))
     find.package(reqlibs)
-->

<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>

# 450k methylation workflow with minfi

Jean-Philippe Fortin [1], Kasper Daniel Hansen [1,2]

[1] Department of Biostatistics, Johns Hopkins Bloomberg School of Public Health, Baltimore, US;

[2] McKusick-Nathans Institute of Genetic Medicine, Johns Hopkins University

```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("BiocStyle")
library("knitr")
options(width=100)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE)
```

# Contents

* [Reading data](#read)
* [Building a RGSet](#construct)

# Introduction

The goal of the tutorial is to present a standard analysis workflow of 450K data with the package `r Biocpkg("minfi")`, incorporating the functions recently added to the package. We invite you to read the software paper recently published \citep{minfi} and the online package vignette on the Bioconductor project \citep{Bioc} for more details. 

We will start from the very beginning by reading input raw data (IDAT files) from an example dataset, and ending with a list of candidate genes for differential methylation. Among others, we will cover quality control assessments, different within-array and between-array normalizations, SNPs removal, sex prediction, differentially methylated positions (DMPs) analysis and bump hunting for differentially methylated regions (DMRs). 

If time permits, we will introduce a complementary interactive visualization tool package available through Bioconductor, `r Biocpkg("shinyMethyl")`, that allows interactive quality control assessment. 

#### 450k Array design and terminology
In this section, we introduce briefly the 450K array as well as the terminology used throughout the `r Biocpkg("minfi")` package. 
Each sample is measured on a single array, in two different color channels (red and green). As the name of the platform indicates, each array measures more than 450,000 CpG positions. For each CpG, we have two measurements: a methylated intensity and an unmethylated intensity. Depending on the probe design, the signals are reported in different colors:

For *Type I* design, both signals are measured in the same color: one probe for the methylated signal and one probe for the unmethylated signal. For *Type II* design, only one probe is used. The intensity read in the green channel measures the methylated signal, and the intensity read in the red channel measures the unmethylated signal.

Two commonly measures are used to report the methylation levels: Beta values and M values. 

**Beta value**:
$$\frac{M}{M + U + 100}$$

where $M$ and $U$ denote the methylated and unmethylated signals respectively. 

**MValue**:
$$Mval = \log{\biggl(\frac{M}{U}\biggr)}$$

**DMP**: Differentially methylated position: single genomic position that has a different methylated level in two different groups of samples (or conditions)

**DMR**: Differentially methylated region: when consecutive genomic locations are differentially methylated in the same direction. 

**Array**: One sample 

**Slide**: Physical slide containing 12 arrays ($6 \times 2$ grid)

**Plate** Physical plate containing at most 8 slides (96 arrays). For this tutorial, we use **batch** and plate interchangeably. 


## Reading Data
The starting point of `r Biocpkg("minfi")` is reading the .IDAT files with the built-in function *read.450k.exp*. Several options are available: the user can specify the sample filenames to be read in along with the directory path, or can specify the directory that contains the files. In the latter case, all the files with the extension .IDAT located in the directory will be loaded into \R{}. The user can also read in a sample sheet, and then use the sample sheet to load the data into a *RGChannelSet*. For more information, see the \texttt{minfi} vignette. Here, we will load the dataset containing 6 samples from the `r Biocpkg("minfiData")` package using the sample sheet provided within the package:

```{r}
library(minfi)
library(minfiData)
baseDir <- system.file("extdata",package="minfiData")
targets <- read.450k.sheet(baseDir)
targets
RGSet <- read.450k.exp(base = baseDir, targets = targets)
```

The class of RGSet is a *RGChannelSet* object. This is the initial object of a minfi analysis that contains the raw intensities in the green and red channels. Note that this object contains the intensities of the internal control probes as well. Because we read the data from a data sheet experiment, the phenotype data is also stored in the *RGChannelSet* and can be accessed via the accessor command *pData*:

```{r}
phenoData <- pData(RGSet)
phenoData[,1:6]
```

The *RGChannelSet* stores also a manifest object that contains the probe design information of the array:

```{r}
manifest <- getManifest(RGSet)
manifest
head(getProbeInfo(manifest))
```

See the [minfi vignette](http://bioconductor.org/packages/release/bioc/vignettes/minfi/inst/doc/minfi.pdf) for more information. 

## Quality control

#### QC plot 

`r Biocpkg("minfi")` provides a simple quality control plot that uses the log median intensity in both the methylated (M) and unmethylated (U) channels. When plotting these two medians against each other, it has been observed that good samples cluster together, while failed samples tend to separate and have lower median intensities. In order to obtain the methylated and unmethylated signals, we need to convert the *RGChannelSet* to an object containing the methylated and unmethylated signals using the function *preprocessRaw*. It takes as input a *RGChannelSet* and converts the red and green intensities to methylated and unmethylated signals according to the special 450K probe design, and returns the converted signals in a new object of class *MethylSet*. It does not perform any normalization.

```{r}
MSet <- preprocessRaw(RGSet) 
MSet
```

The accessors *getMeth* and *getUnmeth* can be used to get the methylated and unmethylated intensities matrices:

```{r}
head(getMeth(MSet)[,1:3])
head(getUnmeth(MSet)[,1:3])
```

The functions *getQC* and *plotQC* are designed to extract and plot the quality control information from the *MethylSet*:

```{r}
qc <- getQC(MSet)
head(qc)
plotQC(qc)
```

Moreover, the function *addQC* applied to the *MethylSet* will add the QC information to the phenotype data. 

To further explore the quality of the samples, it is useful to look at the Beta value densities of the samples, with the option to color the densities by group:

```{r}
densityPlot(MSet, sampGroups = phenoData$Sample_Group)
```

or density bean plots:

```{r}
densityBeanPlot(MSet, sampGroups = phenoData$Sample_Group)
```

`r Biocpkg("shinyMethyl")` is particularly useful to visualize all plots at the same time in an interactive fashion. 

#### Control probes plot

The 450k array contains several internal control probes that can be used to assess the quality control of different sample preparation steps (bisulfite conversion, hybridization, etc.). The values of these control probes are stored in the initial *RGChannelSet* and can be plotted by using the function *controlStripPlot* and by specifying the control probe type:

```{r}
controlStripPlot(RGSet, controls="BISULFITE CONVERSION II")
```

All the plots above can be exported into a pdf file in one step using the function *qcReport*:

```{r}
qcReport(RGSet, pdf= "qcReport.pdf")
```

## MethylSet and RatioSet
A *MethylSet* objects contains the methylated and unmethylated signals, but no methylation summary measures such as Beta values of M values. 

A *RatioSet* object is a class designed to store Beta values and/or M values instead of the methylated and unmethylated signals. An optional copy number matrix, *CN*, the sum of the methylated and unmethylated signals, can be also stored. Mapping a *MethylSet* to a *RatioSet* is irreversible, i.e. one cannot technically retrieve the methylated and unmethylated signals from a *RatioSet*.  A *RatioSet* can be created with the function *ratioConvert*:

```{r}
ratioSet <- ratioConvert(MSet, what = "both", keepCN = TRUE)
ratioSet
```

The functions *getBeta*, *getM* and *getCN* return respectively the Beta value matrix, M value matrix and a the Copy Number matrix. 

```{r}
beta <- getBeta(ratioSet)
```

## GenomicRatioSet 

The function *mapToGenome* applied to a *RatioSet* object will add genomic coordinates to each probe together with some additional annotation information. The output object is a *GenomicRatioSet* (class holding M or/and Beta values together with associated genomic coordinates). It is possible to merge the manifest object with the genomic locations by setting the option *mergeManifest* to *TRUE*.

```{r}
gset <- mapToGenome(ratioSet)
gset
```

Note that the *GenomicRatioSet* extends the class *SummarizedExperiment*. Here are the main accessors functions to access the data:

```{r}
beta <- getBeta(gset)
m <- getM(gset)
cn <- getCN(gset)
```

```{r}
sampleNames <- sampleNames(gset)
probeNames <- featureNames(gset)
pheno <- pData(gset)
```
  
To return the probe locations as a *GenomicRanges* objects, one can use the accessor *granges*:

```{r}
gr <- granges(gset)
head(gr, n= 3)
```

To access the full annotation, one can use the command *getAnnotation*:

```{r}
annotation <- getAnnotation(gset)
names(annotation)
```


## SNPs

Because the presence of SNPs inside the probe body or at the nucleotide extension can have important consequences on the downstream analysis,  `r Biocpkg("minfi")` offers the possibility to remove such probes. The function *getSnpInfo*, applied to a *GenomicRatioSet*, returns a data frame with 6 columns containing the SNP information of the probes:

```{r}
snps <- getSnpInfo(gset)
head(snps,10)
```

*Probe*, *CpG* and *SBE* correspond the SNPs present inside the probe body, at the CpG interrogation and at the single nucleotide extension respectively. The columns with *rs* give the names of the SNPs while the columns with *maf* gives the minor allele frequency of the SNPs based on the dbSnp database. The function *addSnpInfo* will add to the *GenomicRanges* of the *GenomicRatioSet* the 6 columns: 

```{r}
gset <- addSnpInfo(gset)
```

We strongly recommend to drop the probes that contain either a SNP at the CpG interrogation or at the single nucleotide extension.The function *dropLociWithSnps* allows to drop the corresponding probes. Here is an example where we drop the probes containing a SNP at the CpG interrogation and/or at the single nucleotide extension, for any minor allele frequency:


```{r}
gset <- dropLociWithSnps(gset, snps=c("SBE","CpG"), maf=0)
```



## Preprocessing and normalization

So far, we did not use any normalization to process the data. Different normalization procedures are available in `r Biocpkg("minfi")`.

### preprocessRaw

As seen before, it converts a *RGChannelSet* to a *MethylSet* by converting the Red and Green channels into a matrix of methylated signals and a matrix of unmethylated signals. No normalization is performed. 

Input: *RGChannelSet*
Output: *MethylSet*

### preprocessIllumina

Convert a *RGChannelSet* to a *MethylSet* by implementing the preprocessing choices as available in Genome Studio: background subtraction and control normalization. Both of them are optional and turning them off is equivalent to raw preprocessing (*preprocessRaw*):

```{r}
MSet.illumina <- preprocessIllumina(RGSet, bg.correct = TRUE,
                               normalize = "controls")
```

Input: *RGChannelSet*
Output: *MethylSet*

### preprocessSWAN

Perform Subset-quantile within array normalization (SWAN) \citep{swan}, a within-array normalization correction for the technical differences between the Type I and Type II array designs. The algorithm matches the Beta-value distributions of the Type I and Type II probes by applying a within-array quantile normalization separately for different subsets of probes (divided by CpG content). The input of SWAN is a *MethylSet*, and the function returns a *MethylSet* as well. If an *RGChannelSet* is provided instead, the function will first call *preprocessRaw* on the *RGChannelSet*, and then apply the SWAN normalization. We recommend setting a seed (using *set.seed*) before using *preprocessSWAN* to ensure that the normalized intensities will be reproducible. 

```{r}
MSet.swan <- preprocessSWAN(RGSet)
```

Input: *RGChannelSet* or *MethylSet*
Output: *MethylSet*

### preprocessQuantile

\subsection*{preprocessQuantile}
This function implements stratified quantile normalization preprocessing.  The normalization procedure is applied to the Meth and Unmeth intensities separately. The distribution of type I and type II signals is forced to be the same by first quantile normalizing the type II probes across samples and then interpolating a reference distribution to which we normalize the type I probes. Since probe types and probe regions are confounded and we know that DNAm distributions vary across regions we stratify the probes by region before applying this interpolation. Note that this algorithm relies on the assumptions necessary for quantile normalization to be applicable and thus is not recommended for cases where global changes are expected such as in cancer-normal comparisons. Note that this normalization procedure is essentially similar to one previously presented \citep{Touleimat:2012}. The different options can be summarized into the following list:

\begin{itemize}
\item[1)] If *fixMethOutlier* is *TRUE*, the functions fixes outliers of both the methylated and unmethylated channels when small intensities are close to zero. 

\item[2)] If *removeBadSamples* is *TRUE*, it removes bad samples using the QC criterion discussed previously 

\item[3)] Performs stratified subset quantile normalization if *quantileNormalize=TRUE* and *stratified=TRUE*

\item[4)] Predicts the sex (if not provided in the *sex* argument)  using the function *getSex* and normalizes males and females separately for the probes on the X and Y chromosomes 
\end{itemize}

```{r}
gset.quantile <- preprocessQuantile(RGSet, fixOutliers = TRUE,
removeBadSamples = TRUE, badSampleCutoff = 10.5,
quantileNormalize = TRUE, stratified = TRUE, 
mergeManifest = FALSE, sex = NULL)
```

Input: *RGChannelSet*
Output: *GenomicRatioSet*


**Note**: The function returns a *GenomicRatioSet* object ready for downstream analysis. 

### preprocessNoob

```{r}
gset.noob <- preprocessNoob(RGSet, dyeCorr=TRUE)
```


### preprocessFunnorm

The function *preprocessFunnorm* implements the functional normalization algorithm developed in \citep{funnorm}. Briefly, it uses the internal control probes present on the array to infer between-array technical variation. It is particularly useful for studies comparing conditions with known large-scale differences, such as cancer/normal studies, or between-tissue studies. It has been shown that for such studies, functional normalization outperforms other existing approaches \citep{funnorm}. By default, is uses the first two principal components of the control probes to infer the unwanted variation. 

```{r}
gset.funnorm <- preprocessFunnorm(RGSet)
```

Input: *RGChannelSet*
Output: *GenomicRatioSet*

As the *preprocessQuantile* function, it returns a *GenomicRatioSet* object.
 
 
 ## A/B compartments prediction





