<!--
 %\VignetteEngine{knitr::knitr}
-->

---
Author: Jean-Philippe Fortin, Kasper Daniel Hansen
Date: July 13, 2015
---

<!-- rmarkdown v1 -->
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{450k methylation workflow with minfi}
-->

<!-- to compile this: library("knitr"); knit2html("methylation450k.Rmd") -->

<!--
     # a list of all required libraries:
     reqlibs = sub(".*library\\(\"(.*?)\"\\).*","\\1",grep("library\\(",readLines("methylation450k.Rmd"),value=TRUE))
     find.package(reqlibs)
-->

<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>

# 450k methylation analysis workflow with minfi

Jean-Philippe Fortin [1], Kasper Daniel Hansen [1,2]

[1] Department of Biostatistics, Johns Hopkins Bloomberg School of Public Health, Baltimore, US;

[2] McKusick-Nathans Institute of Genetic Medicine, Johns Hopkins University

```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("BiocStyle")
library("knitr")
options(width=100)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE)
```

# Contents

* [Counting reads](#count)
* [Building a DESeqDataSet](#construct)
* [Visual exploration](#eda)
* [Differential expression](#de)
* [Diagnostic plots](#diagnostic)
* [Annotation](#annotate)
* [ReportingTools](#report)
* [Ease of exchanging methods](#methods)

# Introduction

